### DISCUSSION TITLE
Kotlin Multiplatform support

### DISCUSSION THREAD

**Discussion by blakelee** [square/workflow-kotlin#1206] [Show and tell]
I was messing around with workflow to see if I could get it working for a KMP project and I managed to get it working for Compose Multiplatform. I was trying to manage state for a personal app I'm making and after seeing the [existing solutions recommended on jetbrains' website](https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-navigation-routing.html#third-party-alternatives), I decided it was worth a shot to see if I could make it myself. Here is a simple hello world showing it working

<img width="1512" alt="image" src="https://github.com/square/workflow-kotlin/assets/9651986/510d0a08-5f28-45bb-8eeb-ca3b0222daea">

How did I do it? 
1. I noticed that `workflow-core` and `workflow-runtime` already supported KMP.
2. I noticed that @MilanJovic92 [made a post here](https://github.com/square/workflow-kotlin/pull/1185) to move some more code to a module that supports KMP, so I made my own module with just the classes mentioned.
3. The only thing left was the `workflow-ui/compose` module so I imported all of those classes except for `ViewEnvironmentWithComposeSupport.kt` and removed the functions related to `ViewFactory` in `ScreenComposableFactory.kt` which left me with POKO's and nothing Android related.
4. At this point there was a small issue, I needed access to `LocalLifecycleOwner` in `WorkflowRendering.kt` but I needed a KMP solution. There was also [this recent change](https://issuetracker.google.com/issues/328263448) which caused some friction. `LocalLifecycleOwner` was moved to `androidx.lifecycle:lifecycle-runtime-compose` which doesn't have KMP binaries. Luckily for us Jetbrains just released [their own version for KMP here](https://www.jetbrains.com/help/kotlin-multiplatform-dev/compose-lifecycle.html#lifecycle-implementation) which is available in the Compose plugin 1.6.10.
5. Now that I have access to the current lifecycle `compose.ui` in the compose plugin as well as `org.jetbrains.androidx.lifecycle:lifecycle-runtime-compose`, I could edit `WorkflowRendering.kt` to support our KMP lifecycle. Simply import `androidx.lifecycle.compose.LocalLifecycleOwner` then modify the function `rememberChildLifecycleOwner` where the `parentLifecycle` will now be `androidx.compose.ui.platform.LocalLifecycleOwner.current.lifecycle`
6. You now have access to workflow in Compose multiplatform.

I faced a lot of dependency issues trying to figure this out. So as of right now I could only get it working on Android and iOS with Kotlin 2.0.0, Compose-plugin 1.6.10 and Lifecycle-runtime-compose 2.8.0.

If we wanted to make these changes public, we'd have to take the Android code out of `workflow-ui/compose` and possibly make another module for the Android View factories.

**Comment by blakelee** [comment]
Here's a sample project with it running https://github.com/blakelee/WorkflowMultiplatform

**Comment by XhstormR** [comment]
cool
